src/darwin/Framework/CHIP/MTROperationalCredentialsDelegate.mm

```diff
@@ -112,9 +115,171 @@
+        commissioningParameters = mCppCommissioner->GetCommissioningParameters();
+    });
+    VerifyOrReturnError(commissioningParameters.HasValue(), CHIP_ERROR_INCORRECT_STATE);
+
+    // Attestation Elements, nonce and signature will have a value in Commissioning Params as the CSR needs a signature or else we
+    // cannot trust it
+    ReturnErrorOnFailure(
+        chip::Credentials::DeconstructAttestationElements(commissioningParameters.Value().GetAttestationElements().Value(),
+            certificationDeclarationSpan, attestationNonceSpan, timestampDeconstructed, firmwareInfoSpan, vendorReserved));
+
+    AttestationInfo * attestationInfo =
+        [[AttestationInfo alloc] initWithChallenge:AsData(attestationChallenge)
+                                             nonce:AsData(commissioningParameters.Value().GetAttestationNonce().Value())
+                                          elements:AsData(commissioningParameters.Value().GetAttestationElements().Value())
+                                 elementsSignature:AsData(commissioningParameters.Value().GetAttestationSignature().Value())
+                                               dac:AsData(DAC)
+                                               pai:AsData(PAI)
+                          certificationDeclaration:AsData(certificationDeclarationSpan)
+                                      firmwareInfo:AsData(firmwareInfoSpan)];
+
+    dispatch_sync(mNocChainIssuerQueue, ^{
+        [mNocChainIssuer onNOCChainGenerationNeeded:csrInfo
+                                    attestationInfo:attestationInfo
+                       onNOCChainGenerationComplete:^void(NSData * operationalCertificate, NSData * intermediateCertificate,
+                           NSData * rootCertificate, NSData * ipk, NSNumber * adminSubject, NSError * __autoreleasing * error) {
+                           onNOCChainGenerationComplete(
+                               operationalCertificate, intermediateCertificate, rootCertificate, ipk, adminSubject, error);
+                       }];
+    });
+
+    return CHIP_NO_ERROR;
+}
+
+void MTROperationalCredentialsDelegate::setNSError(CHIP_ERROR err, NSError * __autoreleasing * outError)
+{
+    if (outError) {
+        *outError = [MTRError errorForCHIPErrorCode:err];
+    }
+}
+
+void MTROperationalCredentialsDelegate::onNOCChainGenerationComplete(NSData * operationalCertificate,
+    NSData * intermediateCertificate, NSData * rootCertificate, NSData * _Nullable ipk, NSNumber * _Nullable adminSubject,
+    NSError * __autoreleasing * error)
+{
+    if (operationalCertificate == nil || intermediateCertificate == nil || rootCertificate == nil) {
+        setNSError(CHIP_ERROR_INVALID_ARGUMENT, error);
+        return;
+    }
+
+    // use ipk and adminSubject from CommissioningParameters if not passed in.
+    // Dereferencing mCppCommissioner as it would be set to point to a valid Cpp commissioner by now, as we are in the middle of
+    // commissioning
```

Comment:
I'm not sure this true.  We're coming here from some async callback from our API consumer.  The commissioner could have been shut down by now, and then this is a use-after-free.